borsh-rs/borsh/src/schema_helpers.rs

```diff
@@ -0,0 +1,28 @@
+use crate::schema::BorshSchemaContainer;
+use crate::{BorshDeserialize, BorshSchema, BorshSerialize};
+use std::io::{Error, ErrorKind};
+
+/// Deserialize this instance from a slice of bytes, but assume that at the beginning we have
+/// bytes describing the schema of the type. We deserialize this schema and verify that it is
+/// correct.
+pub fn try_from_slice_with_schema<T: BorshDeserialize + BorshSchema>(v: &[u8]) -> Result<T, Error> {
+    let (schema, object) = <(BorshSchemaContainer, T)>::try_from_slice(v)?;
+    if T::schema_container() != schema {
```

Comment:
How expensive to do this? And why do we need to verify it?

Reply:
I would expect it to be faster than the actual deserialization of the schema since it does not do memory allocation. We do it to make sure that when someone tries to deserialize object they are using a correct schema for it, just an additional layer of safety.